test/runtime/mmap.c

```diff
@@ -0,0 +1,533 @@
+};
+
+#define strncat_pretty(str, _cat, len) \
+    if (str[0] != 0) {\
+	strncat(str, "|"_cat, len);\
+    } else {\
+	strncat(str, _cat, len);\
+    }
+
+static void mmap_flags_to_str(char * str, unsigned long str_len, 
+	unsigned long flags)
+{
+    memset(str, 0, str_len);
+
+    if (flags & MAP_ANONYMOUS) {
+	strncat_pretty(str, "MAP_ANONYMOUS", str_len);
+    }
+	
+    if (flags & MAP_PRIVATE) {
+	strncat_pretty(str, "MAP_PRIVATE", str_len);
+    }
+
+    if (flags & MAP_SHARED) {
+	strncat_pretty(str, "MAP_SHARED", str_len);
+    }
+
+    if (flags & MAP_FIXED) {
+	strncat_pretty(str, "MAP_FIXED", str_len);
+    }
+}
+
+static void mmap_test(void)
+{
+    int seed, i;
+
+    printf("** starting mmap tests\n");
+
+    srand(1);
+    printf("  performing sparse_anon_mmap_test with seed=1...\n");
+    fflush(stdout);
+    sparse_anon_mmap_test();
+
+    seed = time(NULL);
+    srand(seed);
+    printf("  performing sparse_anon_mmap_test with seed=%d...\n", seed);
+    fflush(stdout);
+    sparse_anon_mmap_test();
+
+    for (i = 0; i < NR_MMAP_TESTS; i++) {
+	char str[64];
+	void * mmap_addr;
+	unsigned long size;
+
+	if (tests[i].flags & MAP_FIXED)
+	    mmap_addr = (void *)round_up_page(landing_pad); 
+	else
+	    mmap_addr = NULL;
+
+	if (tests[i].flags & MAP_ANONYMOUS)
+	    size = PAGESIZE;
+	else
+	    size = 0;
+
+	mmap_flags_to_str(str, 64, tests[i].flags);
+	printf("  performing mmap_flag_test(%s)...\n", str);
+	fflush(stdout);
+	mmap_flags_test(tests[i].filename, mmap_addr, size, tests[i].flags);
+    }
+
+    printf("  performing munmap test...\n");
+    fflush(stdout);
+    {
+	void * mmap_addr;
+
+	mmap_addr = mmap(NULL, PAGESIZE, PROT_READ, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
+	if (mmap_addr == MAP_FAILED) {
+	    perror("mmap failed");
+	    exit(EXIT_FAILURE);
+	}
+	
+	if (munmap(mmap_addr, PAGESIZE)) {
+	    perror("munmap failed");
+	    exit(EXIT_FAILURE);
+	}
+    }
+
+    printf("** all mmap tests passed\n");
+
+}
+
+static inline bool check_mincore_vec(u8 * vec, int nr_pages)
+{
+    int i;
+    for (i = 0; i < nr_pages; i++) {
+	if (vec[i] != 1)
+	    return false;
+    }
+    return true;
+}
+
+static void
+__mincore(void * addr, unsigned long length, u8 * vec)
+{
+    int ret, nr_pages;
+
+    nr_pages = length >> PAGELOG;
+    
+    ret = mincore(addr, length, vec);
+    if (ret) {
+	perror("mincore failed");
+	exit(EXIT_FAILURE);
+    }
+
+    if (!check_mincore_vec(vec, nr_pages)) {
+	fprintf(stderr, "mincore did not set vector entries\n");
+	exit(EXIT_FAILURE);
+    }
+}
+
+static void mincore_test(void)
+{
+    int ret;
+    u8 * vec;
+    void * addr;
+
+    printf("** starting mincore tests\n");
+    fflush(stdout);
+
+    vec = malloc(sizeof(u8) * 10);
+    if (vec == NULL) {
+	perror("malloc failed");
+	exit(EXIT_FAILURE);
+    }
+
+    /* test something on the stack */
+    printf("  performing mincore on stack address...\n");
+    fflush(stdout);
+    addr = (void *)round_down_page(mincore_test);
```

Comment:
Wouldn't mincore_test be an address on a code page from the program ELF?

Reply:
yep, my mistake